File: frontend/apps/ios/Pantopus/Core/Networking/Models/PrivacyHandshake/PrivacyHandshakeDTOs.swift
//
//  PrivacyHandshakeDTOs.swift
//  Pantopus
//
//  Decoder shapes for the T3.4 handshake flow.
//

import Foundation

/// Envelope from `GET /api/personas/:handle/fan-handle-suggestion`.
public struct FanHandleSuggestionResponse: Decodable, Sendable {
    public let suggestion: String?
    public let locked: Bool?
    public let identity: AudienceIdentityDTO?
}

public struct AudienceIdentityDTO: Decodable, Sendable, Hashable {
    public let id: String
    public let handle: String?
    public let displayName: String?
    public let avatarUrl: String?

    enum CodingKeys: String, CodingKey {
        case id, handle
        case displayName
        case avatarUrl
    }
}

/// Envelope from `GET /api/personas/:id/follow/status`.
public struct FollowStatusResponse: Decodable, Sendable {
    public let following: Bool?
    public let status: String?
    public let relationshipType: String?
    public let notificationLevel: String?
}

/// Envelope from `POST /api/personas/:id/follow` (handshake mode).
/// Tier 1 returns `{follow, status, membership}`; tier > 1 returns
/// `{requiresPayment, subscribeUrl, handshake}`.
public struct HandshakeSubmitResponse: Decodable, Sendable {
    public let follow: HandshakeFollowDTO?
    public let status: String?
    public let membership: HandshakeMembershipDTO?
    public let requiresPayment: Bool?
    public let subscribeUrl: String?
    public let handshake: HandshakeEchoDTO?
}

public struct HandshakeFollowDTO: Decodable, Sendable, Hashable {
    public let id: String?
    public let status: String?
    public let relationshipType: String?
}

public struct HandshakeMembershipDTO: Decodable, Sendable, Hashable {
    public let id: String?
    public let fanHandle: String?
    public let tierId: String?
    public let status: String?

    enum CodingKeys: String, CodingKey {
        case id, status
        case fanHandle = "fan_handle"
        case tierId = "tier_id"
    }
}

public struct HandshakeEchoDTO: Decodable, Sendable, Hashable {
    public let tierRank: Int?
    public let tierId: String?
    public let fanHandle: String?

    enum CodingKeys: String, CodingKey {
        case tierRank = "tier_rank"
        case tierId = "tier_id"
        case fanHandle = "fan_handle"
    }
}

/// Validation error shape returned by `POST /follow` when the body
/// fails Joi validation (e.g. handle pattern). 400 status with a
/// `details` array.
public struct HandshakeValidationErrorDTO: Decodable, Sendable {
    public struct Detail: Decodable, Sendable {
        public let path: String?
        public let message: String?
    }

    public let error: String?
    public let code: String?
    public let details: [Detail]?
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/api/models/handshake/PrivacyHandshakeDtos.kt
@file:Suppress("PackageNaming")

package app.pantopus.android.data.api.models.handshake

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

// GET /api/personas/:handle/fan-handle-suggestion

@JsonClass(generateAdapter = true)
data class FanHandleSuggestionResponse(
    val suggestion: String? = null,
    val locked: Boolean? = null,
    val identity: AudienceIdentityDto? = null,
)

@JsonClass(generateAdapter = true)
data class AudienceIdentityDto(
    val id: String,
    val handle: String? = null,
    val displayName: String? = null,
    val avatarUrl: String? = null,
)

// GET /api/personas/:id/follow/status

@JsonClass(generateAdapter = true)
data class FollowStatusResponse(
    val following: Boolean? = null,
    val status: String? = null,
    val relationshipType: String? = null,
    val notificationLevel: String? = null,
)

// POST /api/personas/:id/follow

/** Request body. `acknowledged_platform_trust` MUST be `true`. */
@JsonClass(generateAdapter = true)
data class HandshakeBody(
    @Json(name = "tier_rank") val tierRank: Int,
    @Json(name = "fan_handle") val fanHandle: String,
    @Json(name = "fan_display_name") val fanDisplayName: String? = null,
    @Json(name = "fan_avatar_url") val fanAvatarUrl: String? = null,
    @Json(name = "acknowledged_platform_trust") val acknowledgedPlatformTrust: Boolean = true,
    @Json(name = "acknowledged_using_pantopus_username") val acknowledgedUsingPantopusUsername: Boolean? = null,
)

@JsonClass(generateAdapter = true)
data class HandshakeSubmitResponse(
    val follow: HandshakeFollowDto? = null,
    val status: String? = null,
    val membership: HandshakeMembershipDto? = null,
    val requiresPayment: Boolean? = null,
    val subscribeUrl: String? = null,
    val handshake: HandshakeEchoDto? = null,
)

@JsonClass(generateAdapter = true)
data class HandshakeFollowDto(
    val id: String? = null,
    val status: String? = null,
    val relationshipType: String? = null,
)

@JsonClass(generateAdapter = true)
data class HandshakeMembershipDto(
    val id: String? = null,
    @Json(name = "fan_handle") val fanHandle: String? = null,
    @Json(name = "tier_id") val tierId: String? = null,
    val status: String? = null,
)

@JsonClass(generateAdapter = true)
data class HandshakeEchoDto(
    @Json(name = "tier_rank") val tierRank: Int? = null,
    @Json(name = "tier_id") val tierId: String? = null,
    @Json(name = "fan_handle") val fanHandle: String? = null,
)

/** 4xx error envelope from the handshake POST. */
@JsonClass(generateAdapter = true)
data class HandshakeValidationErrorDto(
    val error: String? = null,
    val code: String? = null,
    val details: List<HandshakeValidationDetail>? = null,
)

@JsonClass(generateAdapter = true)
data class HandshakeValidationDetail(
    val path: String? = null,
    val message: String? = null,
)

// PATCH /api/personas/:id/follow/preferences

@JsonClass(generateAdapter = true)
data class FollowPreferencesBody(
    @Json(name = "notification_level") val notificationLevel: String? = null,
    @Json(name = "muted_until") val mutedUntil: String? = null,
)

@JsonClass(generateAdapter = true)
data class FollowPreferencesResponse(
    val ok: Boolean? = null,
)
